packages/plugin-react/src/jsx-runtime/babel-restore-jsx.ts

```diff
@@ -0,0 +1,243 @@
+    const [nameNode, propsNode, ...childNodes] = node.arguments
+
+    const name = getJSXName(nameNode)
+    if (name === null) {
+      return null //name is required
+    }
+
+    const props = getJSXProps(propsNode)
+    if (props === null) {
+      return null //no props → [], invalid → null
+    }
+
+    const children = getJSXChildren(childNodes)
+    if (children === null) {
+      return null //no children → [], invalid → null
+    }
+
+    if (
+      t.isJSXMemberExpression(name) &&
+      t.isJSXIdentifier(name.object) &&
+      name.object.name === 'React' &&
+      name.property.name === 'Fragment'
+    ) {
+      return t.jsxFragment(
+        t.jsxOpeningFragment(),
+        t.jsxClosingFragment(),
+        children
+      )
+    }
+
+    // self-closing tag if no children
+    const selfClosing = children.length === 0
+    const startTag = t.jsxOpeningElement(
+      name,
+      props as (t.JSXAttribute | t.JSXSpreadAttribute)[],
+      selfClosing
+    )
+    const endTag = selfClosing ? null : t.jsxClosingElement(name)
+
+    return t.jsxElement(startTag, endTag, children, selfClosing)
+  }
+
+  /**
+   * Get a JSXIdentifier or JSXMemberExpression from a Node of known type.
+   * Returns null if a unknown node type, null or undefined is passed.
+   */
+  function getJSXName(
+    node: t.Node
+  ): t.JSXMemberExpression | t.JSXIdentifier | null {
+    if (node == null) {
+      return null
+    }
+
+    const name = getJSXIdentifier(node)
+    if (name !== null) {
+      return name
+    }
+
+    if (!t.isMemberExpression(node)) {
+      return null
+    }
+    const object = getJSXName(node.object)
+    const property = getJSXName(node.property)
+    if (object === null || property === null) {
+      return null
+    }
+    return t.jsxMemberExpression(object, property as t.JSXIdentifier)
+  }
+
+  /**
+   * Get a array of JSX(Spread)Attribute from a props ObjectExpression.
+   * Handles the _extends Expression babel creates from SpreadElement nodes.
+   * Returns null if a validation error occurs.
+   */
+  function getJSXProps(node: t.Node): t.Node[] | null {
+    if (node == null || isNullLikeNode(node)) {
+      return []
+    }
+
+    if (
+      t.isCallExpression(node) &&
+      t.isIdentifier(node.callee, { name: '_extends' })
+    ) {
+      const props = node.arguments.map(getJSXProps)
+      //if calling this recursively works, flatten.
+      if (props.every((prop) => prop !== null)) {
+        return props as unknown as t.Node[]
+      }
+    }
+
+    if (!t.isObjectExpression(node) && t.isExpression(node))
+      return [t.jsxSpreadAttribute(node)]
+
+    if (!isPlainObjectExpression(node)) {
+      return null
+    }
+    return node.properties.map((prop) =>
+      t.isObjectProperty(prop)
+        ? t.jsxAttribute(
+            getJSXIdentifier(prop.key)!,
+            getJSXAttributeValue(prop.value)
+          )
+        : t.jsxSpreadAttribute((prop as t.SpreadElement).argument)
```

Comment:
This needs to be a `as t.SpreadElement` but could theoretically also be a `t.ObjectMethod`  
We should improve the flow analysis here or maybe fix it?!